File: javascript/breadth_search.js
const { getGraph } = require("./utils");

/**
 * @description - Алгоритм поиска в ширину по графу, сложность O(V+E). V — общее количество вершин. E — общее количество граней (ребер). 
 * Граф - это абстрактная структура данных представляющая собой множество вершин и набор ребер, те соединение м/у парами вершин. Наглядный пример - карта, 
 * где есть города с соединением через маршруты (В нашем случае ребра), а города вершины. Ребра бывают однонаправленными и двунаправленными (A <-> B, A->B). 
 * Для реализации алгоритма следует использовать очередь (queue - первый зашел, первый вышел). *FIFO - First in First
 * @param {Object} graph 
 * @param {string} start 
 * @param {string} end 
 */
function breathSearch(graph, start, end) {
   let queue = []
   queue.push(start)
   while (queue.length > 0) {
      const current = queue.shift()
      if (!graph[current]) {
         graph[current] = []
      }
      if (graph[current].includes(end)) {
         return true
      } else {
         queue = [...queue, ...graph[current]]
      }
   }
   return false
}

console.log(breathSearch(getGraph(), 'a', 'e'))

File: c++/breadth_search.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <vector>
#include "utils.h"

bool breathSearch(const std::unordered_map<char, std::vector<char>>& graph, char start, char end) {
    std::queue<char> queue;
    std::unordered_set<char> visited;

    queue.push(start);

    while (!queue.empty()) {
        char current = queue.front();
        queue.pop();

        if (visited.find(current) == visited.end()) {
            visited.insert(current);

            if (graph.find(current) != graph.end()) {
                for (char neighbor : graph.at(current)) {
                    if (neighbor == end) {
                        return true;
                    }
                    queue.push(neighbor);
                }
            }
        }
    }

    return false;
}

int main() {
    std::unordered_map<char, std::vector<char>> graph = getGraph();
    
    if (breathSearch(graph, 'a', 'e')) {
        std::cout << "Path found" << std::endl;
    } else {
        std::cout << "Path not found" << std::endl;
    }

    return 0;
}